spring-boot/proxy/src/main/java/eu/coatrack/proxy/metrics/MetricsCounterService.java

```diff
@@ -21,131 +21,91 @@
-                    matcher.group(0), serviceApiName, path);
-            if (path == null) {
-                // if api consumer did not specifc path, set as "/"
-                path = "/";
-            } else if (path.endsWith("/") && !path.equals("/")) {
-                // remove trailing "/" as this is the same path from metrics point of view
-                path = path.substring(0, path.length()-1);
-            }
-        } else {
-            log.warn("matcher {} did not match servlet path {}", matcher, request.getServletPath());
-        }
-
-        LocalDate today = LocalDate.now();
-
-        counterService.increment(new StringBuilder()
-                .append(PREFIX) // [0]
-                .append(SEPARATOR)
-                .append(serviceApiName) // [1]
-                .append(SEPARATOR)
-                .append(requestMethod) // [2]
-                .append(SEPARATOR)
-                .append(apiKey) // [3]
-                .append(SEPARATOR)
-                .append(metricType) // [4]
-                .append(SEPARATOR)
-                .append(httpResponseCode) // [5]
-                .append(SEPARATOR)
-                .append(today) // [6]
-                .append(SEPARATOR)
-                .append(path) // [7]
-                .toString());
     }
 
-    public Metric filterAndTransformMetric(org.springframework.boot.actuate.metrics.Metric inputMetric) {
-
-        String[] elements = inputMetric.getName().split(SEPARATOR);
-        Assert.noNullElements(elements);
-
-        // check if this is a custom metric
-        if (elements[0].equals("counter." + PREFIX)) {
-            log.debug("Metrics for transmission: " + inputMetric.getName());
+    private void incrementAndSetCount(Metric metricToTransmit) {
+        String counterId = createCounterIdFromMetric(metricToTransmit);
 
-            Metric outputMetric = new Metric();
-            outputMetric.setRequestMethod(elements[2]);
-            outputMetric.setCount(inputMetric.getValue().intValue());
-            outputMetric.setMetricsCounterSessionID(counterSessionID);
-            outputMetric.setType(MetricType.valueOf(elements[4]));
+        Counter counter = meterRegistry.find(counterId).counter();
+        if (counter == null){
+            counter = meterRegistry.counter(counterId);
+        }
+        counter.increment();
+        int currentCount = (int) counter.count();
+        metricToTransmit.setCount(currentCount);
+    }
 
-            // add a tmp api key object with just the key value, will later be mapped to entity on CoatRack admin side
-            ApiKey tmpApiKey = new ApiKey();
-            tmpApiKey.setKeyValue(elements[3]);
-            outputMetric.setApiKey(tmpApiKey);
+    private String createCounterIdFromMetric(Metric metric) {
+        StringJoiner stringJoiner = new StringJoiner(SEPARATOR);
+        stringJoiner.add(PREFIX)
+                .add(metric.getApiKey().getServiceApiName())
```

Comment:
This is now `null`, probably based on the latest changes?!

Anyway, it seems like the new approach leads to minor changes in the logic that are hard to track down, so I would suggest to at least keep one general principle from previous version: 

1. first generate the counter ID and increment the counter
2. then build the metric to transmit based on the counter

These two steps were completely de-coupled in the previous version, assuring that there are no interdependencies between them. The new version first does half of step 2, then step 1, then step 2, which makes the versions hard to compare. We should discuss this.

Reply:
Done. I fixed the problem as discussed by exclusively using the MetricsHolder as source of metric data. So, the independency of creating 'counterId' and 'metricToTransmit' is granted.